packages/mui-joy/src/TabPanel/TabPanel.tsx

```diff
@@ -83,8 +91,22 @@ const TabPanel = React.forwardRef(function TabPanel(inProps, ref) {
     ownerState,
     className: classes.root,
   });
+  
+  React.useEffect(() => {
+    if (!hidden && !mounted) {
+      setMounted(true);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [hidden]);
 
-  return <TabPanelRoot {...tabPanelRootProps}>{!hidden && children}</TabPanelRoot>;
+  return (
+    <TabPanelRoot
+      {...tabPanelRootProps}
+      sx={{ visibility: hidden ? "hidden" : "visible" }}
+    >
+      {(keepMounted && mounted) || !hidden ? children : null}
```

Comment:
I believe we can drop the `mounted` state altogether. Just leave this:
```suggestion
      {keepMounted || !hidden ? children : null}
```

Reply:
That would mount the children even when the tab panel is hidden, which is undesirable. The login is to mount the children when the tab panel becomes visible, and keep them mounted for further swapping b/w the tabs.

Reply:
I agree with @mnajdova. Also, I feel like the current logic of the TabPanel is wrong. It should hide the whole TabPanel, not the children.

```js
// This looks better and is consistent with other components, e.g. Modal
if (hidden && !keepMounted) {
  return null;
}
return <TabPanelRoot {...tabPanelRootProps}>{children}</TabPanelRoot>;
```

Reply:
hmm makes sense